### ISSUE TITLE
Percolator query not returning document that the query matches

### ISSUE THREAD

**Issue by suresk** [elastic/elasticsearch#28315]
**Bug Report**

**Elasticsearch version:** 6.1.0 (it appears as though the behavior is the same in 6.0.1, too)

**Plugins installed**: None that don't come out of the box.

**JVM version:** 1.8.0_131

**OS version:** macOS Sierra

**Description of the problem including expected versus actual behavior**

Percolator queries are not matching in instances that I expect them to. Playing around with the debugger, if I force it to skip the extract fields step for a query before it saves (which results in query.extraction_result = failed), subsequent attempts to use that percolated query work as expected. So, it seems possible that this is related to something in the path where queries are ruled out before being fully parsed/executed.

I will include a query and a document below - when I index the document and then run the provided query against it, the document is found. When I add the same query to a percolator document and query the user against it, it is not found (unless I use a debugger and force it to not extract terms).

I've had a tough time making sense of some of the code related to the extracted fields/terms, but I'm still trying to see if I can figure out where it is going wrong. Any pointers would be appreciated.

**Steps to reproduce**

The first 3 steps are just meant to show that the percolator query should match by querying the document from the other direction first.

1. Create a user index

```
curl -H "Content-Type: application/json" -XPUT http://localhost:9200/targeting-users -d '{
  "mappings": {
    "user": {
      "properties": {
        "address": {
          "properties": {
            "city": {
              "type": "keyword"
            },
            "country": {
              "type": "keyword"
            },
            "countryId": {
              "type": "long"
            },
            "postalCode": {
              "type": "keyword"
            },
            "subdivision": {
              "type": "keyword"
            },
            "subdivisionId": {
              "type": "long"
            }
          }
        },
        "affiliations": {
          "type": "nested",
          "properties": {
            "expirationDate": {
              "type": "date",
              "format": "yyyy-MM-dd\u0027T\u0027HH:mm:ss.SSS"
            },
            "location": {
              "properties": {
                "blockingBrands": {
                  "type": "long"
                },
                "city": {
                  "type": "keyword"
                },
                "country": {
                  "type": "keyword"
                },
                "countryId": {
                  "type": "long"
                },
                "description": {
                  "type": "keyword"
                },
                "id": {
                  "type": "keyword"
                },
                "indexTime": {
                  "type": "date",
                  "format": "yyyy-MM-dd\u0027T\u0027HH:mm:ss.SSS"
                },
                "locationId": {
                  "type": "long"
                },
                "orgId": {
                  "type": "long"
                },
                "parentId": {
                  "type": "keyword"
                },
                "postalCode": {
                  "type": "keyword"
                },
                "subdivision": {
                  "type": "keyword"
                },
                "subdivisionId": {
                  "type": "long"
                },
                "targetingLabels": {
                  "type": "long"
                },
                "teamId": {
                  "type": "long"
                },
                "unverifiedBrands": {
                  "type": "long"
                },
                "verifiedBrands": {
                  "type": "long"
                }
              }
            },
            "org": {
              "properties": {
                "categories": {
                  "type": "long"
                },
                "indexId": {
                  "type": "keyword"
                },
                "indexTime": {
                  "type": "date",
                  "format": "yyyy-MM-dd\u0027T\u0027HH:mm:ss.SSS"
                },
                "orgId": {
                  "type": "long"
                },
                "orgName": {
                  "type": "keyword"
                },
                "type": {
                  "type": "keyword"
                }
              }
            },
            "status": {
              "type": "keyword"
            },
            "team": {
              "properties": {
                "approvedBrands": {
                  "type": "long"
                },
                "blockedBrands": {
                  "type": "long"
                },
                "blockingBrands": {
                  "type": "long"
                },
                "categories": {
                  "type": "long"
                },
                "id": {
                  "type": "keyword"
                },
                "indexTime": {
                  "type": "date",
                  "format": "yyyy-MM-dd\u0027T\u0027HH:mm:ss.SSS"
                },
                "industries": {
                  "type": "long"
                },
                "manualBrandApproval": {
                  "type": "boolean"
                },
                "memberSegments": {
                  "type": "long"
                },
                "orgId": {
                  "type": "long"
                },
                "parentId": {
                  "type": "keyword"
                },
                "positions": {
                  "type": "long"
                },
                "privateToImplicitTargeting": {
                  "type": "boolean"
                },
                "teamId": {
                  "type": "long"
                },
                "teamName": {
                  "type": "keyword"
                },
                "type": {
                  "type": "keyword"
                }
              }
            }
          }
        },
        "age": {
          "type": "integer"
        },
        "attributes": {
          "type": "nested",
          "properties": {
            "name": {
              "type": "keyword"
            },
            "values": {
              "type": "keyword"
            }
          }
        },
        "categories": {
          "type": "long"
        },
        "expertRanks": {
          "type": "nested",
          "properties": {
            "categoryId": {
              "type": "long"
            },
            "expirationDate": {
              "type": "date",
              "format": "yyyy-MM-dd\u0027T\u0027HH:mm:ss.SSS"
            },
            "rank": {
              "type": "integer"
            }
          }
        },
        "firstName": {
          "type": "keyword"
        },
        "gender": {
          "type": "keyword"
        },
        "indexTime": {
          "type": "date",
          "format": "yyyy-MM-dd\u0027T\u0027HH:mm:ss.SSS"
        },
        "lastLogin": {
          "type": "date",
          "format": "yyyy-MM-dd\u0027T\u0027HH:mm:ss.SSS"
        },
        "lastName": {
          "type": "keyword"
        },
        "recommendedTo": {
          "type": "nested",
          "properties": {
            "orgId": {
              "type": "long"
            },
            "rating": {
              "type": "double"
            }
          }
        },
        "userId": {
          "type": "keyword"
        },
        "userUuid": {
          "type": "keyword"
        },
        "query": {
          "type": "percolator"
        }
      }
    }
  }
}'
```

2. Index the user

```
curl -H "Content-Type: application/json" -XPUT http://localhost:9200/targeting-users/user/1 -d '{
        "userId": 1,
        "userUuid": "XXXXXXXXXXXXXXXXX",
        "firstName": "Test",
        "lastName": "User",
        "age": 20,
        "gender": "MALE",
        "address": {
            "countryId": 232,
            "country": "US",
            "subdivisionId": 4077,
            "subdivision": "UT",
            "postalCode": null,
            "city": null
        },
        "categories": [
            46,
            185
        ],
        "expertRanks": [],
        "recommendedTo": [],
        "affiliations": [
            {
                "org": {
                    "type": "R",
                    "orgId": 1,
                    "orgName": "Test Company",
                    "categories": [
                        49,
                        177
                    ],
                    "indexTime": "2018-01-04T06:00:26.212"
                },
                "team": {
                    "type": "RETAIL",
                    "teamId": 5,
                    "teamName": "Test Team",
                    "orgId": 1,
                    "privateToImplicitTargeting": false,
                    "manualBrandApproval": false,
                    "approvedBrands": [],
                    "blockedBrands": [],
                    "blockingBrands": [],
                    "positions": [],
                    "categories": [],
                    "memberSegments": [],
                    "industries": [],
                    "indexTime": "2018-01-04T06:00:26.212"
                },
                "location": {
                    "locationId": 10,
                    "teamId": 5,
                    "orgId": 1,
                    "description": "Test Location",
                    "countryId": 232,
                    "country": "US",
                    "subdivisionId": 4077,
                    "subdivision": "UT",
                    "postalCode": "84111",
                    "city": "Salt Lake City",
                    "blockingBrands": [
                        2
                    ],
                    "verifiedBrands": [
                        7
                    ],
                    "unverifiedBrands": [
                        55,
                        80
                    ],
                    "targetingLabels": [],
                    "indexTime": "2018-01-04T06:00:26.212"
                },
                "status": "ACTIVE",
                "expirationDate": null
            }
        ],
        "attributes": [],
        "lastLogin": null,
        "indexTime": "2018-01-04T06:00:26.180"
}'
```

3. Run the following query against the user index

```
curl -H "Content-Type: application/json" -XPOST http://localhost:9200/targeting-users/_search -d '{
"query": {
    "bool": {
      "filter": [
        {
          "nested": {
            "query": {
              "terms": {
                "affiliations.team.type": [
                  "RETAIL",
                  "MANUFACTURER",
                  "PARTNER"
                ],
                "boost": 1
              }
            },
            "path": "affiliations",
            "ignore_unmapped": false,
            "score_mode": "none",
            "boost": 1
          }
        },
        {
          "nested": {
            "query": {
              "bool": {
                "should": [
                  {
                    "bool": {
                      "must_not": [
                        {
                          "exists": {
                            "field": "affiliations.expirationDate",
                            "boost": 1
                          }
                        }
                      ],
                      "disable_coord": false,
                      "adjust_pure_negative": true,
                      "boost": 1
                    }
                  },
                  {
                    "range": {
                      "affiliations.expirationDate": {
                        "from": "now",
                        "to": null,
                        "include_lower": true,
                        "include_upper": true,
                        "boost": 1
                      }
                    }
                  }
                ],
                "disable_coord": false,
                "adjust_pure_negative": true,
                "boost": 1
              }
            },
            "path": "affiliations",
            "ignore_unmapped": false,
            "score_mode": "none",
            "boost": 1
          }
        },
        {
          "nested": {
            "query": {
              "terms": {
                "affiliations.status": [
                  "ACTIVE"
                ],
                "boost": 1
              }
            },
            "path": "affiliations",
            "ignore_unmapped": false,
            "score_mode": "none",
            "boost": 1
          }
        },
        {
          "nested": {
            "query": {
              "bool": {
                "filter": [
                  {
                    "terms": {
                      "affiliations.team.type": [
                        "RETAIL",
                        "MANUFACTURER",
                        "PARTNER"
                      ],
                      "boost": 1
                    }
                  }
                ],
                "disable_coord": false,
                "adjust_pure_negative": true,
                "boost": 1
              }
            },
            "path": "affiliations",
            "ignore_unmapped": false,
            "score_mode": "none",
            "boost": 1
          }
        },
        {
          "bool": {
            "should": [
              {
                "nested": {
                  "query": {
                    "terms": {
                      "affiliations.location.countryId": [
                        232
                      ],
                      "boost": 1
                    }
                  },
                  "path": "affiliations",
                  "ignore_unmapped": false,
                  "score_mode": "none",
                  "boost": 1
                }
              }
            ],
            "disable_coord": false,
            "adjust_pure_negative": true,
            "boost": 1
          }
        },
        {
          "bool": {
            "should": [
              {
                "nested": {
                  "query": {
                    "terms": {
                      "affiliations.location.verifiedBrands": [
                        55
                      ],
                      "boost": 1
                    }
                  },
                  "path": "affiliations",
                  "ignore_unmapped": false,
                  "score_mode": "none",
                  "boost": 1
                }
              },
              {
                "nested": {
                  "query": {
                    "terms": {
                      "affiliations.location.unverifiedBrands": [
                        55
                      ],
                      "boost": 1
                    }
                  },
                  "path": "affiliations",
                  "ignore_unmapped": false,
                  "score_mode": "none",
                  "boost": 1
                }
              }
            ],
            "disable_coord": false,
            "adjust_pure_negative": true,
            "boost": 1
          }
        }
      ],
      "must_not": [
        {
          "nested": {
            "query": {
              "bool": {
                "filter": [
                  {
                    "term": {
                      "affiliations.team.manualBrandApproval": {
                        "value": true,
                        "boost": 1
                      }
                    }
                  },
                  {
                    "bool": {
                      "must_not": [
                        {
                          "terms": {
                            "affiliations.team.approvedBrands": [
                              55
                            ],
                            "boost": 1
                          }
                        }
                      ],
                      "disable_coord": false,
                      "adjust_pure_negative": true,
                      "boost": 1
                    }
                  }
                ],
                "disable_coord": false,
                "adjust_pure_negative": true,
                "boost": 1
              }
            },
            "path": "affiliations",
            "ignore_unmapped": false,
            "score_mode": "none",
            "boost": 1
          }
        },
        {
          "nested": {
            "query": {
              "term": {
                "affiliations.team.blockedBrands": {
                  "value": 55,
                  "boost": 1
                }
              }
            },
            "path": "affiliations",
            "ignore_unmapped": false,
            "score_mode": "none",
            "boost": 1
          }
        },
        {
          "nested": {
            "query": {
              "term": {
                "affiliations.team.blockingBrands": {
                  "value": 55,
                  "boost": 1
                }
              }
            },
            "path": "affiliations",
            "ignore_unmapped": false,
            "score_mode": "none",
            "boost": 1
          }
        },
        {
          "nested": {
            "query": {
              "term": {
                "affiliations.location.blockingBrands": {
                  "value": 55,
                  "boost": 1
                }
              }
            },
            "path": "affiliations",
            "ignore_unmapped": false,
            "score_mode": "none",
            "boost": 1
          }
        }
      ],
      "disable_coord": false,
      "adjust_pure_negative": true,
      "boost": 1
    }
  }
}'
```

**Result:** 1 result
**Expected Results:** 1 result

4. Now, create the percolator index

```
curl -H "Content-Type: application/json" -XPUT http://localhost:9200/tag-queries -d '{
  "mappings": {
    "tag": {
      "properties": {
        "address": {
          "properties": {
            "city": {
              "type": "keyword"
            },
            "country": {
              "type": "keyword"
            },
            "countryId": {
              "type": "long"
            },
            "postalCode": {
              "type": "keyword"
            },
            "subdivision": {
              "type": "keyword"
            },
            "subdivisionId": {
              "type": "long"
            }
          }
        },
        "affiliations": {
          "type": "nested",
          "properties": {
            "expirationDate": {
              "type": "date",
              "format": "yyyy-MM-dd\u0027T\u0027HH:mm:ss.SSS"
            },
            "location": {
              "properties": {
                "blockingBrands": {
                  "type": "long"
                },
                "city": {
                  "type": "keyword"
                },
                "country": {
                  "type": "keyword"
                },
                "countryId": {
                  "type": "long"
                },
                "description": {
                  "type": "keyword"
                },
                "id": {
                  "type": "keyword"
                },
                "indexTime": {
                  "type": "date",
                  "format": "yyyy-MM-dd\u0027T\u0027HH:mm:ss.SSS"
                },
                "locationId": {
                  "type": "long"
                },
                "orgId": {
                  "type": "long"
                },
                "parentId": {
                  "type": "keyword"
                },
                "postalCode": {
                  "type": "keyword"
                },
                "subdivision": {
                  "type": "keyword"
                },
                "subdivisionId": {
                  "type": "long"
                },
                "targetingLabels": {
                  "type": "long"
                },
                "teamId": {
                  "type": "long"
                },
                "unverifiedBrands": {
                  "type": "long"
                },
                "verifiedBrands": {
                  "type": "long"
                }
              }
            },
            "org": {
              "properties": {
                "categories": {
                  "type": "long"
                },
                "indexId": {
                  "type": "keyword"
                },
                "indexTime": {
                  "type": "date",
                  "format": "yyyy-MM-dd\u0027T\u0027HH:mm:ss.SSS"
                },
                "orgId": {
                  "type": "long"
                },
                "orgName": {
                  "type": "keyword"
                },
                "type": {
                  "type": "keyword"
                }
              }
            },
            "status": {
              "type": "keyword"
            },
            "team": {
              "properties": {
                "approvedBrands": {
                  "type": "long"
                },
                "blockedBrands": {
                  "type": "long"
                },
                "blockingBrands": {
                  "type": "long"
                },
                "categories": {
                  "type": "long"
                },
                "id": {
                  "type": "keyword"
                },
                "indexTime": {
                  "type": "date",
                  "format": "yyyy-MM-dd\u0027T\u0027HH:mm:ss.SSS"
                },
                "industries": {
                  "type": "long"
                },
                "manualBrandApproval": {
                  "type": "boolean"
                },
                "memberSegments": {
                  "type": "long"
                },
                "orgId": {
                  "type": "long"
                },
                "parentId": {
                  "type": "keyword"
                },
                "positions": {
                  "type": "long"
                },
                "privateToImplicitTargeting": {
                  "type": "boolean"
                },
                "teamId": {
                  "type": "long"
                },
                "teamName": {
                  "type": "keyword"
                },
                "type": {
                  "type": "keyword"
                }
              }
            }
          }
        },
        "age": {
          "type": "integer"
        },
        "attributes": {
          "type": "nested",
          "properties": {
            "name": {
              "type": "keyword"
            },
            "values": {
              "type": "keyword"
            }
          }
        },
        "categories": {
          "type": "long"
        },
        "expertRanks": {
          "type": "nested",
          "properties": {
            "categoryId": {
              "type": "long"
            },
            "expirationDate": {
              "type": "date",
              "format": "yyyy-MM-dd\u0027T\u0027HH:mm:ss.SSS"
            },
            "rank": {
              "type": "integer"
            }
          }
        },
        "firstName": {
          "type": "keyword"
        },
        "gender": {
          "type": "keyword"
        },
        "indexTime": {
          "type": "date",
          "format": "yyyy-MM-dd\u0027T\u0027HH:mm:ss.SSS"
        },
        "lastLogin": {
          "type": "date",
          "format": "yyyy-MM-dd\u0027T\u0027HH:mm:ss.SSS"
        },
        "lastName": {
          "type": "keyword"
        },
        "recommendedTo": {
          "type": "nested",
          "properties": {
            "orgId": {
              "type": "long"
            },
            "rating": {
              "type": "double"
            }
          }
        },
        "userId": {
          "type": "keyword"
        },
        "userUuid": {
          "type": "keyword"
        },
        "query": {
          "type": "percolator"
        }
      }
    }
  }
}'
```

5. Create a percolator record with the query from step 3

```
curl -H "Content-Type: application/json" -XPUT http://localhost:9200/tag-queries/tag/1 -d '{
  "query": {
    "bool": {
      "filter": [
        {
          "nested": {
            "query": {
              "terms": {
                "affiliations.team.type": [
                  "RETAIL",
                  "MANUFACTURER",
                  "PARTNER"
                ],
                "boost": 1
              }
            },
            "path": "affiliations",
            "ignore_unmapped": false,
            "score_mode": "none",
            "boost": 1
          }
        },
        {
          "nested": {
            "query": {
              "bool": {
                "should": [
                  {
                    "bool": {
                      "must_not": [
                        {
                          "exists": {
                            "field": "affiliations.expirationDate",
                            "boost": 1
                          }
                        }
                      ],
                      "disable_coord": false,
                      "adjust_pure_negative": true,
                      "boost": 1
                    }
                  },
                  {
                    "range": {
                      "affiliations.expirationDate": {
                        "from": "now",
                        "to": null,
                        "include_lower": true,
                        "include_upper": true,
                        "boost": 1
                      }
                    }
                  }
                ],
                "disable_coord": false,
                "adjust_pure_negative": true,
                "boost": 1
              }
            },
            "path": "affiliations",
            "ignore_unmapped": false,
            "score_mode": "none",
            "boost": 1
          }
        },
        {
          "nested": {
            "query": {
              "terms": {
                "affiliations.status": [
                  "ACTIVE"
                ],
                "boost": 1
              }
            },
            "path": "affiliations",
            "ignore_unmapped": false,
            "score_mode": "none",
            "boost": 1
          }
        },
        {
          "nested": {
            "query": {
              "bool": {
                "filter": [
                  {
                    "terms": {
                      "affiliations.team.type": [
                        "RETAIL",
                        "MANUFACTURER",
                        "PARTNER"
                      ],
                      "boost": 1
                    }
                  }
                ],
                "disable_coord": false,
                "adjust_pure_negative": true,
                "boost": 1
              }
            },
            "path": "affiliations",
            "ignore_unmapped": false,
            "score_mode": "none",
            "boost": 1
          }
        },
        {
          "bool": {
            "should": [
              {
                "nested": {
                  "query": {
                    "terms": {
                      "affiliations.location.countryId": [
                        232
                      ],
                      "boost": 1
                    }
                  },
                  "path": "affiliations",
                  "ignore_unmapped": false,
                  "score_mode": "none",
                  "boost": 1
                }
              }
            ],
            "disable_coord": false,
            "adjust_pure_negative": true,
            "boost": 1
          }
        },
        {
          "bool": {
            "should": [
              {
                "nested": {
                  "query": {
                    "terms": {
                      "affiliations.location.verifiedBrands": [
                        55
                      ],
                      "boost": 1
                    }
                  },
                  "path": "affiliations",
                  "ignore_unmapped": false,
                  "score_mode": "none",
                  "boost": 1
                }
              },
              {
                "nested": {
                  "query": {
                    "terms": {
                      "affiliations.location.unverifiedBrands": [
                        55
                      ],
                      "boost": 1
                    }
                  },
                  "path": "affiliations",
                  "ignore_unmapped": false,
                  "score_mode": "none",
                  "boost": 1
                }
              }
            ],
            "disable_coord": false,
            "adjust_pure_negative": true,
            "boost": 1
          }
        }
      ],
      "must_not": [
        {
          "nested": {
            "query": {
              "bool": {
                "filter": [
                  {
                    "term": {
                      "affiliations.team.manualBrandApproval": {
                        "value": true,
                        "boost": 1
                      }
                    }
                  },
                  {
                    "bool": {
                      "must_not": [
                        {
                          "terms": {
                            "affiliations.team.approvedBrands": [
                              55
                            ],
                            "boost": 1
                          }
                        }
                      ],
                      "disable_coord": false,
                      "adjust_pure_negative": true,
                      "boost": 1
                    }
                  }
                ],
                "disable_coord": false,
                "adjust_pure_negative": true,
                "boost": 1
              }
            },
            "path": "affiliations",
            "ignore_unmapped": false,
            "score_mode": "none",
            "boost": 1
          }
        },
        {
          "nested": {
            "query": {
              "term": {
                "affiliations.team.blockedBrands": {
                  "value": 55,
                  "boost": 1
                }
              }
            },
            "path": "affiliations",
            "ignore_unmapped": false,
            "score_mode": "none",
            "boost": 1
          }
        },
        {
          "nested": {
            "query": {
              "term": {
                "affiliations.team.blockingBrands": {
                  "value": 55,
                  "boost": 1
                }
              }
            },
            "path": "affiliations",
            "ignore_unmapped": false,
            "score_mode": "none",
            "boost": 1
          }
        },
        {
          "nested": {
            "query": {
              "term": {
                "affiliations.location.blockingBrands": {
                  "value": 55,
                  "boost": 1
                }
              }
            },
            "path": "affiliations",
            "ignore_unmapped": false,
            "score_mode": "none",
            "boost": 1
          }
        }
      ],
      "disable_coord": false,
      "adjust_pure_negative": true,
      "boost": 1
    }
  }
}'
```

6. Use the user from step 2 to do a percolator query

```
curl -H "Content-Type: application/json" -XPOST http://localhost:9200/tag-queries/_search -d '{
  "query": {
    "percolate": {
      "field": "query",
      "document": {
        "userId": 1,
        "userUuid": "XXXXXXXXXXXXXXXXX",
        "firstName": "Test",
        "lastName": "User",
        "age": 20,
        "gender": "MALE",
        "address": {
          "countryId": 232,
          "country": "US",
          "subdivisionId": 4077,
          "subdivision": "UT",
          "postalCode": null,
          "city": null
        },
        "categories": [
          46,
          185
        ],
        "expertRanks": [],
        "recommendedTo": [],
        "affiliations": [
          {
            "org": {
              "type": "R",
              "orgId": 1,
              "orgName": "Test Company",
              "categories": [
                49,
                177
              ],
              "indexTime": "2018-01-04T06:00:26.212"
            },
            "team": {
              "type": "RETAIL",
              "teamId": 5,
              "teamName": "Test Team",
              "orgId": 1,
              "privateToImplicitTargeting": false,
              "manualBrandApproval": false,
              "approvedBrands": [],
              "blockedBrands": [],
              "blockingBrands": [],
              "positions": [],
              "categories": [],
              "memberSegments": [],
              "industries": [],
              "indexTime": "2018-01-04T06:00:26.212"
            },
            "location": {
              "locationId": 10,
              "teamId": 5,
              "orgId": 1,
              "description": "Test Location",
              "countryId": 232,
              "country": "US",
              "subdivisionId": 4077,
              "subdivision": "UT",
              "postalCode": "84111",
              "city": "Salt Lake City",
              "blockingBrands": [
                2
              ],
              "verifiedBrands": [
                7
              ],
              "unverifiedBrands": [
                55,
                80
              ],
              "targetingLabels": [],
              "indexTime": "2018-01-04T06:00:26.212"
            },
            "status": "ACTIVE",
            "expirationDate": null
          }
        ],
        "attributes": [],
        "lastLogin": null,
        "indexTime": "2018-01-04T06:00:26.180"
      }
    }
  }
}'
```

**Result:** 0 results
**Expected Result:** 1 result

**Comment by martijnvg** [comment]
Hey @suresk, I'm looking at the last 3 snippets here and in step 5 you're trying to store a percolator query with a `percolate` query. I'm not sure what you expect in step 6, but I don't expect that to match with the percolator query you stored in step 5. What would you expect in this case? A `percolate` query is used to matched stored percolator queries and not meant to be used inside a percolator query.

**Comment by suresk** [comment]
Hi @martijnvg - Sorry! It was just a copy/paste error. I had the right description for that step, but grabbed the wrong JSON block. I've updated my issue - does it make more sense now?

**Comment by martijnvg** [comment]
@suresk I see what you mean, but the percolator query in step 5 hasn't been updated yet. Just to be sure that percolator should look like:

```
PUT /tag-queries/tag/1
{
  "query": {
    "bool": {
      "filter": [
        {
          "nested": {
            "query": {
              "terms": {
                "affiliations.team.type": [
                  "RETAIL",
                  "MANUFACTURER",
                  "PARTNER"
                ],
                "boost": 1
              }
            },
            "path": "affiliations",
            "ignore_unmapped": false,
            "score_mode": "none",
            "boost": 1
          }
        },
        {
          "nested": {
            "query": {
              "bool": {
                "should": [
                  {
                    "bool": {
                      "must_not": [
                        {
                          "exists": {
                            "field": "affiliations.expirationDate",
                            "boost": 1
                          }
                        }
                      ],
                      "disable_coord": false,
                      "adjust_pure_negative": true,
                      "boost": 1
                    }
                  },
                  {
                    "range": {
                      "affiliations.expirationDate": {
                        "from": "now",
                        "to": null,
                        "include_lower": true,
                        "include_upper": true,
                        "boost": 1
                      }
                    }
                  }
                ],
                "disable_coord": false,
                "adjust_pure_negative": true,
                "boost": 1
              }
            },
            "path": "affiliations",
            "ignore_unmapped": false,
            "score_mode": "none",
            "boost": 1
          }
        },
        {
          "nested": {
            "query": {
              "terms": {
                "affiliations.status": [
                  "ACTIVE"
                ],
                "boost": 1
              }
            },
            "path": "affiliations",
            "ignore_unmapped": false,
            "score_mode": "none",
            "boost": 1
          }
        },
        {
          "nested": {
            "query": {
              "bool": {
                "filter": [
                  {
                    "terms": {
                      "affiliations.team.type": [
                        "RETAIL",
                        "MANUFACTURER",
                        "PARTNER"
                      ],
                      "boost": 1
                    }
                  }
                ],
                "disable_coord": false,
                "adjust_pure_negative": true,
                "boost": 1
              }
            },
            "path": "affiliations",
            "ignore_unmapped": false,
            "score_mode": "none",
            "boost": 1
          }
        },
        {
          "bool": {
            "should": [
              {
                "nested": {
                  "query": {
                    "terms": {
                      "affiliations.location.countryId": [
                        232
                      ],
                      "boost": 1
                    }
                  },
                  "path": "affiliations",
                  "ignore_unmapped": false,
                  "score_mode": "none",
                  "boost": 1
                }
              }
            ],
            "disable_coord": false,
            "adjust_pure_negative": true,
            "boost": 1
          }
        },
        {
          "bool": {
            "should": [
              {
                "nested": {
                  "query": {
                    "terms": {
                      "affiliations.location.verifiedBrands": [
                        55
                      ],
                      "boost": 1
                    }
                  },
                  "path": "affiliations",
                  "ignore_unmapped": false,
                  "score_mode": "none",
                  "boost": 1
                }
              },
              {
                "nested": {
                  "query": {
                    "terms": {
                      "affiliations.location.unverifiedBrands": [
                        55
                      ],
                      "boost": 1
                    }
                  },
                  "path": "affiliations",
                  "ignore_unmapped": false,
                  "score_mode": "none",
                  "boost": 1
                }
              }
            ],
            "disable_coord": false,
            "adjust_pure_negative": true,
            "boost": 1
          }
        }
      ],
      "must_not": [
        {
          "nested": {
            "query": {
              "bool": {
                "filter": [
                  {
                    "term": {
                      "affiliations.team.manualBrandApproval": {
                        "value": true,
                        "boost": 1
                      }
                    }
                  },
                  {
                    "bool": {
                      "must_not": [
                        {
                          "terms": {
                            "affiliations.team.approvedBrands": [
                              55
                            ],
                            "boost": 1
                          }
                        }
                      ],
                      "disable_coord": false,
                      "adjust_pure_negative": true,
                      "boost": 1
                    }
                  }
                ],
                "disable_coord": false,
                "adjust_pure_negative": true,
                "boost": 1
              }
            },
            "path": "affiliations",
            "ignore_unmapped": false,
            "score_mode": "none",
            "boost": 1
          }
        },
        {
          "nested": {
            "query": {
              "term": {
                "affiliations.team.blockedBrands": {
                  "value": 55,
                  "boost": 1
                }
              }
            },
            "path": "affiliations",
            "ignore_unmapped": false,
            "score_mode": "none",
            "boost": 1
          }
        },
        {
          "nested": {
            "query": {
              "term": {
                "affiliations.team.blockingBrands": {
                  "value": 55,
                  "boost": 1
                }
              }
            },
            "path": "affiliations",
            "ignore_unmapped": false,
            "score_mode": "none",
            "boost": 1
          }
        },
        {
          "nested": {
            "query": {
              "term": {
                "affiliations.location.blockingBrands": {
                  "value": 55,
                  "boost": 1
                }
              }
            },
            "path": "affiliations",
            "ignore_unmapped": false,
            "score_mode": "none",
            "boost": 1
          }
        }
      ],
      "disable_coord": false,
      "adjust_pure_negative": true,
      "boost": 1
    }
  }
}
```

If that is the case then I can reproduce what you mean.

**Comment by suresk** [comment]
Yes, that is what I mean.. I'm not sure why it didn't update before, but it looks like it is updated now.

**Comment by martijnvg** [comment]
I investigated this bug and this caused by the fact that minimum should matches are extracted incorrectly from nested queries. I will work on a fix. @suresk thanks for reporting!

**Comment by leonardocaldas** [comment]
Hey guys, I'm facing a similar problem with nested queries and percolator.

Just to be clear, this problem happens with any nested query, right? Because I have a simple percolate nested query that when executed against the index, it finds the document, but when used with the percolate query, no query is found for the document.

I'll leave my scenario here, so you guys can confirm it.

Elasticsearch version: 5.5.0
Docker Image: docker.elastic.co/elasticsearch/elasticsearch:5.5.0

Creating the index
```
PUT my_index
{
  "mappings": {
    "percolator": {
      "properties": {
        "query": {
          "type": "percolator"
        }
      }
    },
    "my_type": {
      "properties": {
        "tags": {
          "type": "nested",
          "properties": {
            "tag": {
              "type": "keyword"
            }
          }
        }
      }
    }
  }
}
```
Adding a new document
```
POST my_index/my_type/1
{
  "tags": [
    {"tag": "first_tag"},
    {"tag": "second_tag"}
  ]
}
```

Testing the query we're gonna save in the percolator type
```
GET my_index/my_type/_search
{
  "query": {
    "nested": {
      "path": "tags",
      "query": {
        "bool": {
          "must": [
            {
              "match": {
                "tags.tag": "first_tag"
              }
            }
          ]
        }
      }
    }
  }
}
```
The query works as expected.

Adding the query to the percolator type
```
PUT my_index/percolator/1?refresh
{
  "query": {
    "nested": {
      "path": "tags",
      "query": {
        "bool": {
          "must": [
            {
              "match": {
                "tags.tag": "first_tag"
              }
            }
          ]
        }
      }
    }
  }
}
```
Percolating the document
```
GET my_index/_search
{
  "query": {
    "percolate": {
      "field": "query",
      "document_type": "percolator",
      "document": {
        "tags": [
          {
            "tag": "first_tag"
          },
          {
            "tag": "second_tag"
          }
        ]
      }
    }
  }
}
```
Here nothing is found!

I really hope to be doing something wrong, because for me this is a very important feature and my app depends on it.

**Comment by leonardocaldas** [comment]
Hey guys!

After running some more tests I realized what I was doing wrong. I confused the parameter "document_type" in the percolate query, thinking that I should inform the type which holds the queries, when in fact I should inform the type which holds the documents.

So please disregard my previous comment.

Thanks!

**Comment by martijnvg** [comment]
@leonardocaldas Thanks for testing some more. You are right about the fact that this bug can occur  with other queries too. Basically if a percolator query contains duplicated clauses this error can occur. The fix in #28353 will prevent this bug from happening.

**Comment by maricn** [comment]
I've just noticed this bug with `must` condition. I'm executing the following query:
```
{
  "query": {
    "bool": {
      "must": [
        {
          "nested": {
            "query": {
              "bool": {
                "must": [
                  {
                    "term": {
                      "stories.storyId": 2
                    }
                  }
                ]
              }
            },
            "path": "stories"
          }
        },
        {
          "nested": {
            "query": {
              "bool": {
                "must": [
                  {
                    "term": {
                      "stories.storyId": 1
                    }
                  },
                  {
                    "term": {
                      "stories.storyName": "story1"
                    }
                  }
                ]
              }
            },
            "path": "stories"
          }
        },
        {
          "nested": {
            "query": {
              "bool": {
                "must": [
                  {
                    "range": {
                      "stories.executedAt": {
                        "gte": "now-1w"
                      }
                    }
                  }
                ]
              }
            },
            "path": "stories"
          }
        }
      ]
    }
  }
}
```

I get back the document: 
```

{
  "took": 3,
  "timed_out": false,
  "_shards": {
    "total": 5,
    "successful": 5,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": 1,
    "max_score": 4.568616,
    "hits": [
      {
        "_index": "user-180306",
        "_type": "user",
        "_id": "kpF7-mEBH3oCPpbkErLB",
        "_score": 4.568616,
        "_source": {
          "stories": [
            {
              "storyId": 1,
              "storyName": "story1",
              "executedAt": "2018-03-06T07:45:00.000+01:00"
            },
            {
              "storyId": 2,
              "storyName": "story2",
              "executedAt": "2018-03-06T07:55:00.000+01:00"
            }
          ],
          "status": {
            "atHome": true,
            "atWork": false,
            "isRunning": false,
            "isActive": true,
            "isSleeping": false
          }
        }
      }
    ]
  }
}
```

But when i percolate that search, and run that document from result against the percolator, I don't get the percolated query in result list (i get other valid queries).
